overlord/snapstate/autorefresh_gating.go

```diff
@@ -32,6 +35,265 @@ import (
+	}
+	inf, err := snapst.CurrentInfo()
+	if err != nil {
+		return time.Time{}, err
+	}
+	fst, err := os.Stat(inf.MountFile())
+	if err != nil {
+		return time.Time{}, err
+	}
+	return fst.ModTime(), nil
+}
+
+type holdInfo struct {
+	// FirstHeld keeps the time when the given snap was first held for refresh by a gating snap.
+	FirstHeld time.Time `json:"first-held"`
+	// HoldUntil stores the desired end time for holding.
+	HoldUntil time.Time `json:"hold-until"`
+}
+
+func refreshGating(st *state.State) (map[string]map[string]*holdInfo, error) {
+	// held snaps -> holding snap(s) -> first-held/hold-until time
+	var gating map[string]map[string]*holdInfo
+	err := st.Get("snaps-hold", &gating)
+	if err != nil && err != state.ErrNoState {
+		return nil, fmt.Errorf("internal error: cannot get snaps-hold: %v", err)
+	}
+	if err == state.ErrNoState {
+		return make(map[string]map[string]*holdInfo), nil
+	}
+	return gating, nil
+}
+
+// HoldError contains the details of snaps that cannot to be held.
+type HoldError struct {
+	SnapsInError map[string]error
+}
+
+func (h *HoldError) Error() string {
+	l := []string{""}
+	for _, e := range h.SnapsInError {
+		l = append(l, e.Error())
+	}
+	return fmt.Sprintf("cannot hold some snaps:%s", strings.Join(l, "\n - "))
+}
+
+func maxAllowedPostponement(gatingSnap, affectedSnap string) time.Duration {
+	if affectedSnap == gatingSnap {
+		return maxPostponement
+	}
+	return maxOtherHoldDuration
+}
+
+// HoldRefresh marks affectingSnaps as held for refresh for up to holdTime.
+// HoldTime of zero denotes maximum allowed hold time.
+// Holding may fail for only some snaps in which case HoldError is returned and
+// it contains the details of failed ones.
+func HoldRefresh(st *state.State, gatingSnap string, holdDuration time.Duration, affectingSnaps ...string) error {
+	gating, err := refreshGating(st)
+	if err != nil {
+		return err
+	}
+	herr := &HoldError{
+		SnapsInError: make(map[string]error),
+	}
+	now := timeNow()
+	for _, affecting := range affectingSnaps {
+		refreshed, err := lastRefreshed(st, affecting)
+		if err != nil {
+			return err
+		}
+		if _, ok := gating[affecting]; !ok {
+			gating[affecting] = make(map[string]*holdInfo)
+		}
+		hold, ok := gating[affecting][gatingSnap]
+		if !ok {
+			hold = &holdInfo{
+				FirstHeld: now,
+			}
+		}
+
+		defaultDur := maxAllowedPostponement(gatingSnap, affecting)
+		dur := holdDuration
+		if dur == 0 {
+			// duration not specified, using a default one
+			dur = defaultDur
+		} else {
+			// explicit hold duration requested
+			if dur > defaultDur {
+				herr.SnapsInError[affecting] = fmt.Errorf("requested holding duration %s exceeds maximum holding for snap %q", holdDuration, affecting)
+				continue
+			}
+		}
+
+		newHold := now.Add(dur)
+		if newHold.Before(refreshed.Add(maxPostponement)) {
+			hold.HoldUntil = newHold
+			gating[affecting][gatingSnap] = hold
+		} else {
+			// default duration requested but using the maximum exceeds max
```

Comment:
I would move this comment inside the `if` below, otherwise it seems it applies to the whole `else` branch.